sharding/mainchain/smc_client_test.go

```diff
@@ -1,6 +1,192 @@
+	"github.com/ethereum/go-ethereum/sharding/contracts"
+)
+
+var (
+	key, _                   = crypto.HexToECDSA("b71c71a67e1177ad4e901695e1b4b9ee17ae16c6668d313eac2f96dbcda3f291")
+	addr                     = crypto.PubkeyToAddress(key.PublicKey)
+	accountBalance1001Eth, _ = new(big.Int).SetString("1001000000000000000000", 10)
+)
 
 // Verifies that SMCCLient implements the sharding Service inteface.
 var _ = sharding.Service(&SMCClient{})
+
+// mockClient is struct to implement the smcClient methods for testing.
+type mockClient struct {
+	smc         *contracts.SMC
+	depositFlag bool
+	t           *testing.T
+	backend     *backends.SimulatedBackend
+	blockNumber *big.Int
+}
+
+// Mirrors the function in the main file, but instead of having a client to perform rpc calls
+// it is replaced by the simulated backend.
+func (m *mockClient) WaitForTransaction(ctx context.Context, hash common.Hash, durationInSeconds int64) error {
+
+	var receipt *types.Receipt
+
+	ctxTimeout, cancel := context.WithTimeout(ctx, time.Duration(durationInSeconds)*time.Second)
+
+	for err := error(nil); receipt == nil; receipt, err = m.backend.TransactionReceipt(ctxTimeout, hash) {
+
+		if err != nil {
+			cancel()
+			return fmt.Errorf("unable to retrieve transaction: %v", err)
+		}
+		if ctxTimeout.Err() != nil {
+			cancel()
+			return fmt.Errorf("transaction timed out, transaction was not able to be mined in the duration: %v", ctxTimeout.Err())
+		}
+	}
+	cancel()
+	ctxTimeout.Done()
+	log.Info(fmt.Sprintf("Transaction: %s has been mined", hash.Hex()))
+	return nil
+}
+
+// Creates and send Fake Transactions to the backend to be mined, takes in the context and
+// the current blocknumber as an argument and returns the signed transaction after it has been sent.
+func (m *mockClient) CreateAndSendFakeTx(ctx context.Context) (*types.Transaction, error) {
+	tx := types.NewTransaction(m.blockNumber.Uint64(), common.HexToAddress("0x"), nil, 50000, nil, nil)
+	signedtx, err := types.SignTx(tx, types.MakeSigner(&params.ChainConfig{}, m.blockNumber), key)
+	if err != nil {
+		return nil, err
+	}
+	err = m.backend.SendTransaction(ctx, signedtx)
+	if err != nil {
+		return nil, fmt.Errorf("unable to send transaction: %v", err)
+	}
+	return signedtx, nil
+}
+
+func (m *mockClient) Commit() {
+	m.backend.Commit()
+	m.blockNumber = big.NewInt(m.blockNumber.Int64() + 1)
+}
+
+func setup() *backends.SimulatedBackend {
+	backend := backends.NewSimulatedBackend(core.GenesisAlloc{addr: {Balance: accountBalance1001Eth}})
+	return backend
+}
+
+// TestWaitForTransaction tests the WaitForTransaction function in the smcClient, however since for testing we do not have
+// an inmemory rpc server to interact with the simulated backend, we have to define the function in the test rather than
+// in `smc_client.go`.
+// TODO: Test the function in the main file instead of defining it here if a rpc server for testing can be
+// implemented.
+func TestWaitForTransaction(t *testing.T) {
```

Comment:
It looks like you are testing 3 different scenarios. 
Can you split into 3 tests?

TestWaitForTransaction_IsMinedImmediately
TestWaitForTransaction_TimesOut
TestWaitForTransaction_IsCancelledWhenParentCtxCancelled